src/clientiface.h

```diff
@@ -372,7 +373,17 @@ class RemoteClient
 		List of block positions.
 		No MapBlock* is stored here because the blocks can get deleted.
 	*/
-	std::set<v3s16> m_blocks_sent;
+	std::unordered_set<v3s16> m_blocks_sent;
+
+	/*
+		Cache of blocks that have been occlusion culled partitioned by distance/layer.
+		Occlusion culling is expensive.
+		Reset when the player moves or changes view-direction.
+		Outer layers are reset when blocks in inner layers that were visible to the client
+		have been changed, as they could have occluded farther blocks.
+	 */
+	std::vector<std::unordered_set<v3s16>> m_blocks_occ;
```

Comment:
What are usual values of `d_max`? Asking due to the performance of creating many sets.

Reply:
d_max is always <= wanted range which is 63 for viewing_range of 1000. So relatively small.
Usually it is just 2 bigger than the last iteration.